Progpow integration

holiman opened the pull request:
This PR is a test to see how easy it would be to integrate progpow into go-ethereum. It adds a `progpowBlock` to the chain config, making it easy to spin up/join a network with hashimoto->progouw transition, to see how well it plays with other implementations. 

For example, it would be interesting to see if this 
- works well on both big-endian and little-endian systems, and 
- if there are any quirks around fast or light sync
- works with remote miners

Progpow turned out to be extremely easy to add, since it was not a new consensus engine per se, but rather an internal switchover from hashimoto to progpow inside ethash. 

cc @OhGodAGirl

holiman commented:
@eosclassicteam yup, fixed

holiman commented:
> will launch a public testnet with progpow maybe tomorrow

If you do, please make it use hashimoto for a while, so there's a proper switchover _after_ the difficulty has set, and blocktimes are ~14s. Then we would also see how the switch affects blocktimes (though, it's a pretty moot datapoint, since it's CPU-mining and not necessarily reflects performance differences for GPU mining)

karalabe commented:
It would probably make sense to integrate it into puppeth, because then we have **everything** for a testnet and don't have to hack all the pieces together individually.

eosclassicteam commented:
@holiman Yes I would like to configure a new testnet like this

```
TestnetChainConfig = &ChainConfig{
		ChainID:             big.NewInt(3),
		HomesteadBlock:      big.NewInt(0),
		DAOForkBlock:        nil,
		DAOForkSupport:      true,
		EIP150Block:         big.NewInt(0),
		EIP150Hash:          common.HexToHash("0x41941023680923e0fe4d74a34bdac8141f2540e3ae90623718e47d66d1ca4a2d"),
		EIP155Block:         big.NewInt(0),
		EIP158Block:         big.NewInt(0),
		ByzantiumBlock:      big.NewInt(0),
		ConstantinopleBlock: nil,
                ProgpowBlock:      big.NewInt(10000),
		Ethash:              new(EthashConfig),
	}

```

holiman commented:
done! I also added Constantinople to puppeth, seems we hadn't done so yet

karalabe commented:
Do we have any benchmark numbers? I'm curious how much work we're putting on ourselves wrt progpow verification.

holiman commented:
[user@work ethash]$ git diff
```diff
diff --git a/consensus/ethash/algorithm_test.go b/consensus/ethash/algorithm_test.go
index f51d2e2a6..e342eab40 100644
--- a/consensus/ethash/algorithm_test.go
+++ b/consensus/ethash/algorithm_test.go
@@ -864,6 +864,19 @@ func BenchmarkHashimotoLight(b *testing.B) {
        }
 }
 
+// Benchmarks the light verification performance.
+func BenchmarkProgpowLight(b *testing.B) {
+       cache := make([]uint32, cacheSize(1)/4)
+       generateCache(cache, 0, make([]byte, 32))
+
+       hash := hexutil.MustDecode("0xc9149cc0386e689d789a1c2f3d5d169a61a6218ed30e74414dc736e442ef3d1f")
+
+       b.ResetTimer()
+       for i := 0; i < b.N; i++ {
+               progpowLight(datasetSize(1), cache, hash, 0, 0)
+       }
+}
+
 // Benchmarks the full (small) verification performance.
 func BenchmarkHashimotoFullSmall(b *testing.B) {
        cache := make([]uint32, 65536/4)
```
```
[user@work ethash]$ go test -run - -bench Benchmar.*Light
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/consensus/ethash
BenchmarkHashimotoLight-2   	     500	   3930131 ns/op
BenchmarkProgpowLight-2     	      20	  79983262 ns/op
PASS
ok  	github.com/ethereum/go-ethereum/consensus/ethash	7.281s
```
Quite a big diff

holiman commented:
Same for full:
```
[user@work ethash]$ go test -run - -bench Benchmar.*FullSmall
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/consensus/ethash
BenchmarkHashimotoFullSmall-2   	  300000	      5684 ns/op
BenchmarkProgpowFullSmall-2     	    2000	    711026 ns/op
PASS
ok  	github.com/ethereum/go-ethereum/consensus/ethash	4.171s

```

karalabe commented:
The full is not too relevant because (1. geth doesn't cpu mine; 2. testnet cpu mining is fine independent of the speed; 3. it's up to gpus to max the speed). The light is more problematic because that's used during sync. 

Ethash was 4ms according to your benchmark, so that's 6372574 * 4ms = ~7 hours to fully verify mainnet on a single core.

ProgPoW at 80ms would mean 6372574 * 80ms = ~6 days. That's quite a bump.

AlexeyAkhunov commented:
Going from 4ms to 80ms will also affect block propagation quite a lot. It will increase uncle rate, therefore, ostensibly, reduce miner's profitability, and increase the variability of their income.

holiman commented:
Yup. I'm wondering, though, if some low  level optimizations could bring it down an order of magnitude, or if it's inherent in the algo.

The code contains e.g. zeroing a newly alloc:ed byte array. That in itself is not a big timesink, but indicates a lack of familiarity with golang.

holiman commented:
Would be interesting to see benchmarks comparing hashimoto vs progpow on cpp

OhGodAGirl commented:
ProgPoW is tunable - we did not tune it for light-client verification (hence its brutal nature because it is a copy of the GPU code) and left this for the main development team of each client, since each will have their own implementation requirements.

AlexeyAkhunov commented:
@holiman I agree - some optimisation is possible (for example, implementing "clz" and "popcount" functions more efficiently, with tricks from "Hacker's Delight" or Knuth's Volume 4 fascicle 1 "Bitwise tricks"). However, we need to first figure out why it went to 80ms, what is the main contributor

AlexeyAkhunov commented:
I tried some optimisations (reducing allocations, passing arrays in/out functions, pop count, leading zeros) https://gist.github.com/AlexeyAkhunov/b3a5046fd2fa062b7d4c3fadc9194607, but it is still stays at 80ms, unfortunately

AlexeyAkhunov commented:
And, most of the time is spent in the calls to the "generateDatasetItem" function. Each invocation of "progpowLoop" calls it 2*32 = 64 times. The function "progpowLoop" itself is called 64 times inside the "progpow" function, so "generateDatasetItem" is called 64*64 = 4096 times this way. Apart from the the "progpow" function calls "generateDatasetItem" directly 2*4096 = 8192 times. In total, "progpow" calls "generateDatasetItem" 4096+8192 = 12288 times. Hashimoto, on the other hand, cals "generateDatasetItem" only 64*2 = 128 times. My calculations might be off by a factor, but the general idea should be clear - it simply does too much access. Perhaps that was a mistake?

eosclassicteam commented:
@AlexeyAkhunov Perhaps ProgPOW's design is poor for verification, I hope it's not, we can still find a way to implement it 😂

AlexeyAkhunov commented:
@eosclassicteam Efficient verification is one of the important design requirement for a good PoW algorithm. Perhaps it has not been considered. But if the verification performance is so poor, ProgPow hinders too many things - light clients, block propagation (block gas limit will have to be reduced to support that), sync performance - it will take much longer to sync. Everything will get much slower

ifdefelse commented:
Hi all.

The ProgPoW team is looking into this- the CPU verification side is being copied from our GPU mining code, which was required for our own internal testing, but obviously not required for proper client implementation. We'll look into this.

MoneroCrusher commented:
@AlexeyAkhunov @karalabe @holiman according to ProgPOW devs it's a bug that is being addressed. Real verification time will be 8ms, not 80ms. Therefore not a problem.

Arachnid commented:
> @AlexeyAkhunov @karalabe @holiman according to ProgPOW devs it's a bug that is being addressed. Real verification time will be 8ms, not 80ms. Therefore not a problem.

Please let them speak for themselves. They're here on the thread; trying to talk for them just confuses matters.

MoneroCrusher commented:
> > @AlexeyAkhunov @karalabe @holiman according to ProgPOW devs it's a bug that is being addressed. Real verification time will be 8ms, not 80ms. Therefore not a problem.
> 
> Please let them speak for themselves. They're here on the thread; trying to talk for them just confuses matters.

I mainly made that post so it is publicly posted and that the public knows that ~8ms (of course it depends on the CPU and how many cores are utilized) is the real number after the bug is eliminated, not 80ms (I did not know this until 1 hour ago). ProgPOW devs have said they'd rather not engage in too much public discussions anymore. I expect that the people I tagged above know this already, but tagged them anyway in case they didn't. It's important that the public stays up to date too on latest developments as not everyone is part of the discord group(s).
Of course I'm not stopping the ProgPOW devs from propagating that info themselves.

OhGodABird commented:
https://github.com/ifdefelse/go-ethereum/commits/master

There is a slight performance improvement at [ifdefelse/go-ethereum](https://github.com/ifdefelse/go-ethereum/commits/master) mainly reducing the number of generateDatasetItem calls (lookups), producing the following results:
```
$ go test -run - -bench Benchmark.*Light
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/consensus/ethash
BenchmarkHashimotoLight-8                   1000           1565141 ns/op
BenchmarkProgpowLight-8                      100          16306560 ns/op
BenchmarkProgpowOptimalLight-8               300           5343173 ns/op
PASS
ok      github.com/ethereum/go-ethereum/consensus/ethash        10.158s
```

BenchmarkProgpowLight tests the drop-in replacement, while BenchmarkProgpowOptimalLight pregenerates some data which is only implemented in the test right now.

There is still room for improvement, for example all the math can be done as SSE vectors instead of iterative loops. A well-optimized implementation should end up being about just 2x slower than the ethash verification as there is twice as much data read per iteration.

ifdefelse commented:
There is a slight performance improvement at [ifdefelse/go-ethereum](https://github.com/ifdefelse/go-ethereum/commits/master) mainly reducing the number of generateDatasetItem calls (lookups), producing the following results:
```
$ go test -run - -bench Benchmark.*Light
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/consensus/ethash
BenchmarkHashimotoLight-8                   1000           1565141 ns/op
BenchmarkProgpowLight-8                      100          16306560 ns/op
BenchmarkProgpowOptimalLight-8               300           5343173 ns/op
PASS
ok      github.com/ethereum/go-ethereum/consensus/ethash        10.158s
```

BenchmarkProgpowLight tests the drop-in replacement, while BenchmarkProgpowOptimalLight pregenerates some data which is only implemented in the test right now.

There is still room for improvement, for example all the math can be done as SSE vectors instead of iterative loops. A well-optimized implementation should end up being about just 2x slower than the ethash verification as there is twice as much data read per iteration.

eosclassicteam commented:
@ifdefelse Yes I think this is a remarkable change well done 👍

eosclassicteam commented:
@holiman Can we cherry-pick this commit on this PR? https://github.com/ifdefelse/go-ethereum/commit/580566e4dbce9937893d20647356c21c4c19c697

@ifdefelse Also you should change your commit message to ethash: ~~~ next time,

since it is meant to mark package name see https://github.com/ethereum/go-ethereum/blob/master/.github/CONTRIBUTING.md#coding-guidelines

holiman commented:
Yes, I'll probably cherry-pick the relevant commit. I want to implement the optimized version directly, so I'll have to look into the pregeneration a bit. Will be on a plane today, so probably won't have the time to fix this until maybe tomorrow. 

Something that would help, @ifdefelse, would be more testvectors for the verification, so we can ensure that it's working correctly if we do some more optimizations.

AlexeyAkhunov commented:
@ifdefelse I have looked at your code, thank you. I saw what you did there and it makes sense.

HiddeHoogland commented:
@ifdefelse would you be joining the Ethereum Core Devs Meeting 47 friday?

chfast commented on consensus/ethash/progpow.go:
```diff
@@ -0,0 +1,410 @@
+	0x0000800a, 0x8000000a, 0x80008081, 0x00008080, 0x80000001, 0x80008008}
+
+func keccakF800Round(st [25]uint32, r int) [25]uint32 {
+	var keccakfROTC = [24]uint32{1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 2,
```
These values are incorrect, probably copied from Keccak-f[1600]. You cannot rol 32-bit integer by 36 bits.

AlexeyAkhunov replied:
I noticed those two. But it never occurred to me that the were copied from keccak1600. In the implementation (and in Go library math/bits), rolling by 36 bits means rolling by 4 bits :)

chfast replied:
In C this is undefined behavior, but somehow it works.

holiman commented:
I have now implemented the changes to the algo, as well as

* Added an RLP chain containing 1000+ blocks, where the transition from hashimoto to progpow occurs at block 5, 
* Made the block import utilise pregenerated cDag lookup

It's still a bit flaky, perhaps, and I'm not totally sure that the cDag generation is correctly implemented. It works well for `geth ... import`, but it's not production ready. 

Most recent benchmarks:

```
BenchmarkHashimotoLight-4        	     500	   4074180 ns/op
BenchmarkProgpowLight-4          	      50	  25323537 ns/op
BenchmarkProgpowOptimalLight-4   	     200	   8914564 ns/op
```
So `4ms` for hashimoto, `9ms` for progpow, which is more in line with the theoretical estimations.

holiman commented:
Here's another dump of the blocks, for other implementations to compare against: https://gist.github.com/holiman/487b85b53dd59d5b0da14737f6417755

chfast commented on consensus/ethash/progpow.go:
```diff
@@ -0,0 +1,418 @@
+
+	st[8] = lower32(nonce)
+	st[9] = higher32(nonce)
+	for i := 0; i < 8; i++ {
```
Yes, in implementation I've seen the `result` has 128 bits (4 x uint32). Not sure how it works here.

holiman replied:
The `result` is defined as `8 x 32` bits: `make([]uint32, 8)` at
https://github.com/ethereum/go-ethereum/blob/7abf9078c99607b699c43164803c18ba3a41ae40/consensus/ethash/progpow.go#L383

holiman commented:
Ping @OhGodAGirl -- please see the questions from @andresilva and @chfast

fjl commented on consensus/ethash/progpow.go:
```diff
@@ -0,0 +1,418 @@
+		}
+	}
+	return count
+}
```
You can use `bits.OnesCount32` (from math/bits) instead, it's faster.

fjl commented on consensus/ethash/progpow.go:
```diff
@@ -0,0 +1,418 @@
+		}
+	}
+	return uint32(32)
+}
```
You can use `bits.LeadingZeros32` instead.

chfast commented on consensus/ethash/algorithm.go:
```diff
@@ -207,6 +207,27 @@ func generateCache(dest []uint32, epoch uint64, seed []byte) {
 
+// generateCDag generates the cDag used for progpow. If the 'cDag' is nil, this method is a no-op. Otherwise
+// it expects the cDag to be of size progpowCacheWords
+func generateCDag(cDag, cache []uint32, epoch uint64) {
```
Where is the CDag specified?

MoneroCrusher commented:
Lots of code being discussed that I have no idea about, however, is ProgPOW looking good so far? Is an integration determined & planned or are you still evaluating?

HiddeHoogland commented:
> Lots of code being discussed that I have no idea about, however, is ProgPOW looking good so far? Is an integration determined & planned or are you still evaluating?

Same question from me! Looking forward to the Constantinople update. Also interested in the progress of ProgPow.

Daft-Wullie commented:
@AlexeyAkhunov @holiman @chfast @Souptacular 
Just wanted to leave this here for increased visibility, since i remember people asked for some profiler data/GPU utilization stats in one of the recent eth core devs meetings.
In case anyone missed it:
https://medium.com/@ifdefelse/understanding-progpow-performance-and-tuning-d72713898db3
Sorry for the ping and have a great Devcon4.

chrisfranko commented:
Lots of good work going on here at a level that I’ll admit is a little over my head. Keep up the good work everyone!

hackmod commented:
I guess you already know. spec updated at https://github.com/ifdefelse/ProgPOW/commit/a3f62349a1513f0393524683f9671cfe17cca895

holiman commented:
@hackmod yes, I just finished implementing them

hackmod commented on consensus/ethash/progpow_test.go:
```diff
@@ -38,6 +38,56 @@ func TestRandomMerge(t *testing.T) {
+
+		}
+		if exp := expCdag4080_to_4095[i]; exp != cDag[4080+i] {
+			t.Error("test %d (+4080), exp %x != %x", i, exp, cDag[4080+i])
```
`s/t.Error/t.Errorf/`

holiman replied:
Thanks!

holiman commented:
My machines shows about `4ms` versus `10ms`. The latest changes increased the progpow verification time a bit, since it needs to access (and thus generate) more dag items during verification.

holiman commented:
Yes, you're probably right @chfast. Just to clarify what I meant -- it did _not_  increase from `4ms` to `10ms` with the latest change, but ethash was `4ms`, and progpow now `10ms` , and earlier it was `8ms`. 

Anyhow, there's probably room for improvements.

holiman commented:
Also, rebased on master

naikmyeong commented:
https://github.com/ethereum/EIPs/pull/1588

hackmod commented:
works nicely with the latest ProgPoW version of ethminer (slightly modified version)
(current ProgPoW-ethminer does not have it's own CPU routines
please see also https://github.com/ifdefelse/ProgPOW/issues/14 )

![image](https://user-images.githubusercontent.com/32324335/48650927-cdb59680-ea3b-11e8-9ede-750a38a73411.png)

hackmod commented:
(I guess you already know.)
ProgPoW spec v0.9.2 updated! https://github.com/ifdefelse/ProgPOW/commit/824cd791634204c4cc7e31f84bb76c0c84895bd3

holiman commented:
Yes, I'm aware, but thanks even so!

holiman commented:
Updated to 0.9.2

naikmyeong commented:
@holiman Cah you rebase the whole progpow branch with master so we can solve the conflict?

holiman commented:
I'll wait until Constantinople block is merged, otherwise I'll have to redo it again

holiman commented:
Ok, it's merged. I'll fix in a bit

naikmyeong commented on eth/backend.go:
```diff
@@ -244,6 +244,7 @@ func CreateConsensusEngine(ctx *node.ServiceContext, chainConfig *params.ChainCo
 			DatasetDir:     config.DatasetDir,
 			DatasetsInMem:  config.DatasetsInMem,
 			DatasetsOnDisk: config.DatasetsOnDisk,
+			ProgpowBlockNumber: chainConfig.ProgpowBlock,
```
Need lint here 😂

naikmyeong commented:
Some errors on ethash.go

`consensus/ethash/ethash.go:753:5:warning: should merge variable declaration with assignment on next line (S1021) (gosimple)`

@holiman

naikmyeong commented:
@holiman Build failed 😂 

```
# github.com/ethereum/go-ethereum/tests
tests/block_test_util.go:119:28: not enough arguments in call to ethash.NewShared
	have ()
	want (*big.Int)
```

cryptozeny commented on consensus/ethash/ethash.go:
```diff
@@ -564,7 +573,12 @@ func NewFullFaker() *Ethash {
+func NewShared(progpowNumber *big.Int) *Ethash {
+	ethashMu.Lock()
+	if sharedEthash == nil {
+		sharedEthash = New(Config{"", 3, 0, "", 1, 0, ModeNormal, progpowNumber}, nil, false)
```
For a clearer explanation, I would like to:

```go
func NewShared(progpowNumber *big.Int) *Ethash {
	ethashMu.Lock()
	if sharedEthash == nil {
		sharedEthash = New(
			Config{
				CacheDir: 		"",
				CachesInMem:  	3,
				CachesOnDisk: 	0,
				"",
				DatasetsInMem:  1,
				DatasetsOnDisk: 0,
				ModeNormal,
				progpowNumber,
			},
			nil, false,
		)
	}
	ethashMu.Unlock()
	return &Ethash{shared: sharedEthash}
}
```

My code maybe wrong... :sweat_smile: Could you fix like this?

holiman replied:
If we do it like that, and ever modify `Config` to add a field, that field will be left to default. As it is right now, it will cause a compilation error, which is probably better.

MoneroCrusher commented:
Is there a reason for `PROGPOW_PERIOD = 50`
According to a conversation with @SChernykh it takes 0.2-0.3 s per GPU to compile the random program.
So why not reduce the block period further? Even 1 block would be possible. Could take 5 to be safe in case of incredible miner luck. Anything I'm missing?

ifdefelse commented:
> Is there a reason for `PROGPOW_PERIOD = 50`?
> According to a conversation with @SChernykh it takes 0.2-0.3 s per GPU to compile the random program.
> So why not reduce the block period further? Even 1 block would be possible. Could take 5 to be safe in case of incredible miner luck. Anything I'm missing?

That's probably an off-topic question for this thread focusing on integration to geth, but I'll provide a quick answer.

A lot of mining systems use extremely slow processors that could take multiple seconds to compile the random kernel.  Setting PERIOD=1 would likely be too much overhead for those systems.  Once per epoch (30,000 blocks or 5 days) would be too infrequent, losing much of the benefit of random programs.  Anything in the range of 10 to 250 blocks would have worked in practice.  50 blocks (aka 12.5 minutes) was picked as a reasonable tradeoff point.

SChernykh commented:
> A lot of mining systems use extremely slow processors that could take multiple seconds to compile the random kernel.

0.2-0.3 s per GPU if for low-end Intel Celeron CPU, if the random kernel is compiled separately from other code. It can also be pre-compiled in background since it only depends on block height, so it's not an argument.

MoneroCrusher commented:
> That's probably an off-topic question for this thread focusing on integration to geth, but I'll provide a quick answer.
> 
> A lot of mining systems use extremely slow processors that could take multiple seconds to compile the random kernel. Setting PERIOD=1 would likely be too much overhead for those systems. Once per epoch (30,000 blocks or 5 days) would be too infrequent, losing much of the benefit of random programs. Anything in the range of 10 to 250 blocks would have worked in practice. 50 blocks (aka 12.5 minutes) was picked as a reasonable tradeoff point.

As SChernykh has pointed out again, a very cheap and bad Intel Celeron G3930 that you can likely find in many mining setups has a compilation time of <0.3s per GPU and can be pre-compiled in the background.
So if PROGPOW_PERIOD=2 was chosen I don't see how it would affect even the worst system out there if it can be pre-compiled in the background as SChernykh has said. But I do see potential disadvantages to specialized hardware, especially FPGAs that would have to have ultra-hyperscale-fast bitstream loading that would shave away valuable mining time.

I don't know if any of my points are valid since I'm not a dev. But if i.e. PROGPOW_PERIOD=2 truly had no effect on your typical GPU mining setup hashrate, then I don't see any disadvantages of choosing such a parameter as it would increase random programs 25-fold while not hurting anything.

holiman commented:
This is now rebased on master. This implementation does _not_ include progpow v `0.9.3` changes, suggested here: https://medium.com/@ifdefelse/progpow-progress-da5bb31a651b. 
Although, at this point, neither does the parity implementation at the moment: https://github.com/paritytech/parity-ethereum/pull/9762 
 
cc @chfast @andresilva LMK if we should upgrade this

chfast commented:
I think it should be merged with 0.9.2. The 0.9.3 is a proposal what we can then compare with 0.9.2 by having an another PR.

holiman commented:
I agree. @karalabe @fjl PTAL

karalabe commented on cmd/puppeth/wizard_genesis.go:
```diff
@@ -229,6 +229,10 @@ func (w *wizard) manageGenesis() {
+		fmt.Println()
+		fmt.Printf("Which block should ProgPow come into effect? (default = %v)\n", w.conf.Genesis.Config.ProgpowBlock)
+		w.conf.Genesis.Config.ProgpowBlock = w.readDefaultBigInt(w.conf.Genesis.Config.ProgpowBlock)
+
```
We should only do this for `ethash` based networks. For Clique, we might as well skip.

karalabe commented on cmd/puppeth/module_dashboard.go:
```diff
@@ -633,6 +633,7 @@ func deployDashboard(client *sshClient, network string, conf *config, config *da
 		"Byzantium":         conf.Genesis.Config.ByzantiumBlock,
 		"Constantinople":    conf.Genesis.Config.ConstantinopleBlock,
 		"ConstantinopleFix": conf.Genesis.Config.PetersburgBlock,
+		"ProgPow":           conf.Genesis.Config.ProgpowBlock,
```
Nit, can we call this ProgPoW?

karalabe commented on cmd/utils/flags.go:
```diff
@@ -1573,12 +1573,13 @@ func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chai
+				DatasetDir:         stack.ResolvePath(eth.DefaultConfig.Ethash.DatasetDir),
+				DatasetsInMem:      eth.DefaultConfig.Ethash.DatasetsInMem,
+				DatasetsOnDisk:     eth.DefaultConfig.Ethash.DatasetsOnDisk,
+				ProgpowBlockNumber: config.ProgpowBlock,
```
ProgpowBlockNumber -> ProgpowBlock

karalabe commented on consensus/ethash/algorithm_test.go:
```diff
@@ -772,6 +772,21 @@ func BenchmarkHashimotoLight(b *testing.B) {
 	}
 }
 
+// BenchmarkProgpowLight Benchmarks the light verification performance (not counting cDag generation).
```
Benchmarks -> benchmarks

karalabe commented on consensus/ethash/ethash.go:
```diff
@@ -564,7 +574,12 @@ func NewFullFaker() *Ethash {
 // NewShared creates a full sized ethash PoW shared between all requesters running
 // in the same process.
-func NewShared() *Ethash {
+func NewShared(progpowNumber *big.Int) *Ethash {
```
This doesn't really make sense if the progpow number can be changed. The point of the shared ethash engine is that it *can* be shared by every tester thread. The new code just races to who can allocate the fastest, but if multiple progpow blocksnumbers are used, it will blow up either way.

What you could do to fix it is to make the sharedEthash a `map[uint64]*Ethash`, then you can have multiple shared ones, one for each progpow transition period.

karalabe commented on consensus/ethash/progpow.go:
```diff
@@ -0,0 +1,414 @@
+package ethash
```
Copyright header pls.

karalabe commented on tests/progpow_testvectors.json:
```diff
@@ -0,0 +1,38 @@
+  [59999, "f5c50ba5c0d6210ddb16250ec3efda178de857b2b1703d8d5403bd0f848e19cf", "02edb6275bd221e3",
+    "653eda37d337e39d311d22be9bbd3458d3abee4e643bee4a7280a6d08106ef98",
+    "341562d10d4afb706ec2c8d5537cb0c810de02b4ebb0a0eea5ae335af6fb2e88"]
+]
```
The tests folder should only contain the consensus tests. Please move package specific tests into the respective folders (e.g. package/testdata). That's how all our packages are implemented, makes things self-contained instead of referencing some global path.

holiman commented:
Addresed the concerns, ptal @karalabe

holiman commented:
Rebased on master, squashed most commits

holiman commented:
Rebased, and afaik up to date with https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1057.md at `0.9.3`. cc @gcolvin

holiman commented on consensus/ethash/algorithm.go:
```diff
@@ -207,6 +207,27 @@ func generateCache(dest []uint32, epoch uint64, seed []byte) {
+	}
+
+	elapsed := time.Since(start)
+	log.Info("Generated progpow cDag", "elapsed", common.PrettyDuration(elapsed), "epoch", epoch)
```
```suggestion
	log.Debug("Generated progpow cDag", "elapsed", common.PrettyDuration(elapsed), "epoch", epoch)
```
